### PR TITLE
SpdxDocumentReporter: Implement mapping to the `extracted license info` list

### PR DESCRIPTION
**Pull request by fviernau**
See individual commit messages for the details.

### CODE REVIEW

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt [hunk 1]
```diff
@@ -211,6 +215,25 @@ private fun ScanResult.toSpdxPackageVerificationCode(): SpdxPackageVerificationC
...
     )
 
+private fun SpdxDocument.addExtractedLicenseInfo(licenseTextProvider: LicenseTextProvider): SpdxDocument {
+    val nonSpdxLicenses = packages.flatMapTo(mutableSetOf()) {
+        SpdxExpression.parse(it.licenseConcluded).licenses() + SpdxExpression.parse(it.licenseDeclared).licenses()
+    }.filter {
+        !SpdxConstants.isNotPresent(it) && SpdxLicense.forId(it) == null && SpdxLicenseException.forId(it) == null
+    }
+
+    val extractedLicenseInfo = nonSpdxLicenses.sorted().mapNotNull { license ->
```
Comment (sschuberth) [inline reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:225]: Hint: If you're sorting `nonSpdxLicenses` here, you could also map to a `sortedSetOf()` above instead.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt [hunk 2]
```diff
@@ -211,6 +215,25 @@ private fun ScanResult.toSpdxPackageVerificationCode(): SpdxPackageVerificationC
         packageVerificationCodeValue = summary.packageVerificationCode.toLowerCase()
     )
 
+private fun SpdxDocument.addExtractedLicenseInfo(licenseTextProvider: LicenseTextProvider): SpdxDocument {
+    val nonSpdxLicenses = packages.flatMapTo(mutableSetOf()) {
+        SpdxExpression.parse(it.licenseConcluded).licenses() + SpdxExpression.parse(it.licenseDeclared).licenses()
```
Comment (sschuberth) [inline reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: Hmm, is it really correct to use concluded *and* declared licenses here? At least in ORT's model, concluded trumps declared (and detected), so I would have expected to use concluded here, and only fall back to declared if there is no concluded license.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt [hunk 3]
```diff
@@ -211,6 +215,25 @@ private fun ScanResult.toSpdxPackageVerificationCode(): SpdxPackageVerificationC
...
+    }.filter {
+        !SpdxConstants.isNotPresent(it) && SpdxLicense.forId(it) == null && SpdxLicenseException.forId(it) == null
```
Comment (sschuberth) [inline reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:222]: Nit: That initial double negation is not so super-ice. Should be spend an `isPresent()` convenience function in `SpdxConstants`?

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 2 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: I'm not following your argumentation why the fact that concluded trumps declared in ORT should imply that the license texts for the declared licenses do not need to be provided in an SpdxDocument (for packages with a concluded license).

Anyhow the SPDX specification says "One instance shall be created for every unique license or licensing information being referenced that does not match one of the licenses on the SPDX License List.", see https://github.com/spdx/spdx-spec/blob/development/v2.2.1/chapters/composition-of-an-SPDX-document.md#525-other-licensing-information-detected-.

I think the implementation `as-is` matches that specification and should not be changed. Do you agree @sschuberth ?

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 1 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:225]: I'd like to leave it as is.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 3 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:222]: Done.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 2 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: Ah, to me the important bit to quote is "This section is used for any detected, declared or concluded licenses that are NOT on the SPDX License List". So aren't we then missing detected licenses here?

Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: Well, that's a different topic. But yes, detected licenses are missing. They are intentionally not yet implemented and will be added in a future iteration as it needs further clarifications.

Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: @sschuberth  does it resolve this discussion?

Comment (sschuberth) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: As it is completely not obvious from neither the commit message nor the code that you're aware that the implementation is not complete, please change the commit message to

```
The extracted license info list is supposed to contain "any detected,
declared or concluded licenses that are NOT on the SPDX License List".
Only implement this for concluded and declared licenses for now as
also getting detected licenses requires more work.
```

and add

```
// TODO: Also add detected non-SPDX licenses here.
```

before line 220 to the code.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 3 above)
Comment (sschuberth) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:222]: I guess this was not pushed yet?

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 2 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt:220]: Done.

File: reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt (hunk 3 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/SpdxDocumentModelMapper.kt]: Now it is.

**Pull request merged**